src/python/WMCore/JobSplitting/EventAwareLumiBased.py

```diff
@@ -309,3 +316,29 @@ def algorithm(self, *args, **kwargs):
         self.lumiChecker.closeJob(self.currentJob)
         self.lumiChecker.fixInputFiles()
         return
+
+    def getFilesSortedByLocation(self, eventsPerJob):
+        """
+        _getFilesSortedByLocation_
+
+        Retrieves a list of files available and sort them by location.
+        If the fileset is closed, resume the splitting. Otherwise check whether
```

Comment:
What do you do if the file set is not closed but there are not enough events? Is there a timeout? Is success guaranteed at some point?

Reply:
That's a good question. As mentioned in my first comment, I didn't add any timeout because that means I'd have to query every single file to be splitted and compare their insert time against the timeout value.
If there are not enough events, it will get merged once the fileset gets closed, which is when the last job for that task completes. I've just run further tests and they confirm it works.

Reply:
Do we have to worry about 12 or 24 hour cleanup of un-merged space?

Reply:
First, this job splitter should never run on unmerged input. Second, even our default merge job splitters don't have timeouts. If you have a processing workflow with two jobs and the two jobs finish two weeks apart, the output from the first job will just sit on unmerged space for two weeks (assuming they are too small to be elevated to merged immediately). WMAgent has always worked that way. I do have timeouts for Repack and RepackMerge, but that's mostly because people are super-paranoid about RAW (plus we have/had problems with the StorageManager bookkeeping that sometimes kept a Repack workflow in limbo for weeks).

Reply:
Just a correction, these splitters can run on unmerged data for some specific taskchain setup (KeepOutput=False)

Reply:
Ok, unmerged cleanup time is 2 weeks for most sites. "Normally" that should not be an issue but in case of workflow tails we definitely do run the risk of loosing these files.

Problem is, adding a safety timeout of lets say a week requires to run a more expensive data discovery query for the job splitter. Inert time isn't available from the wmbs_file_details table, you need to join with wmbs_fileset_files. That join requires another join with wmbs_subscription. Neither of these two joins are very expensive, so I think the performance implications should be minimal.

Reply:
Thing is, if this is really a problem Ops should be seeing this already. Maybe they are just working around it ? Worth asking them and see how much these worries are justified ?

Reply:
what could be the problem for Ops ? loosing unmerged files because site would be deleting files after 2 weeks ? the actual policy is not 2 weeks, but "2 weeks and not in the list of protected directories" which will protect any deletion of the unmerged output of on-going workflows. Unless sites don't play by the rule, we are safe.
If the timeout of 2 weeks is on collecting enough work for the subsequent job, will this impact the completion time significantly ?

Reply:
It can impact in the case there is a last job or two going into cooloff over and over. The change proposed relies on:
1. enough events or lumis for at least one job, properly splitted
2. when the fileset is closed, so all the input files were made available

and not on a time limit for files available for splitting. That would slightly increase the complexity and load on the database.
Perhaps a better approach would be making an in-memory cache for subscriptions and first time a file was made available., if we deem necessary to have it. Otherwise this patch is good as is.